tick/preprocessing/longitudinal_features_product.py

```diff
@@ -0,0 +1,236 @@
+             'short', not %s" %exposure_type)
+        self.n_jobs = n_jobs
+        self.exposure_type = exposure_type
+        self.reset()
+
+    def reset(self):
+        self._set("_n_init_features", None)
+        self._set("_n_output_features", None)
+        self._set("_n_intervals", None)
+        self._set("_mapper", {})
+        self._set("_preprocessor", None)
+        self._set("_fitted", False)
+
+    @property
+    def mapper(self):
+        """Get the mapping between the feature products and column indexes.
+
+        Returns
+        -------
+        output : `dict`
+            The column index - feature mapping.
+        """
+        if not self._fitted:
+            raise ValueError("cannot get mapper if object has not been fitted.")
+        return self._mapper.copy()
+
+    def fit(self, X):
+        """Fit the feature product using the features matrices list.
+
+        Parameters
+        ----------
+        X : `[np.ndarrays]`  or `[csr_matrices]`, shape=(n_intervals, n_features)
+            The list of features matrices.
+
+        Returns
+        -------
+        output : `LongitudinalFeaturesProduct`
+            The fitted current instance.
+        """
+        self.reset()
+        # TODO: check input type ?
+        # TODO: Check that all the matrices have the same size ?
+        n_intervals, n_init_features = X[0].shape
+        if n_init_features < 2:
+            raise ValueError("There should at least two features to compute\
+                     product features.")
+        self._set("_n_init_features", n_init_features)
+        self._set("_n_intervals", n_intervals)
+        comb_it = combinations(range(n_init_features), 2)
+        mapper = {i + n_init_features: c for i, c in enumerate(comb_it)}
+        self._set("_mapper", mapper)
+        self._set("_n_output_features", int(n_init_features +
+                                            comb(n_init_features, 2)))
+
+        if sps.issparse(X[0]) and self.exposure_type == "infinite":
+            self._set("_preprocessor", SparseLongitudinalFeaturesProduct(X))
+
+        self._set("_fitted", True)
+
+        return self
+
+    def transform(self, X):
+        """Add the product features to the given features matrices list.
+
+        Parameters
+        ----------
+        X : `[np.ndarrays]`  or `[csr_matrices]`, shape=(n_intervals, n_features)
+            The list of features matrices.
+
+        Returns
+        -------
+        output : `[np.ndarrays]`  or `[csr_matrices]`, shape=(n_intervals, n_features)
+            The list of features matrices with added product features.
+        """
+
+        # TODO: check matrix dimensions and consistency with self.fit ?
+        if self.exposure_type == "short":
+            X_with_products = self.short_exposure_products(X)
+        elif self.exposure_type == "infinite":
+            X_with_products = self.infinite_exposure_products(X)
```

Comment:
else raise a ValueError or something like that (in case a user plays with set._set()